# AOL Reduced Dataset (Citizen Lab)

In [1]:
import pandas
from collections import Counter

## Treated datasets

In [2]:
%%time
# Loads AOL-treated dataset into data DataFrame.
data = pandas.read_csv('AOL-treated.csv', low_memory=False, on_bad_lines='warn', index_col=0)

CPU times: user 23.2 s, sys: 1.83 s, total: 25 s
Wall time: 24.8 s


In [3]:
%%time
# Loads domain match for AOL-treated dataset and Citizen Lab test list classification into topics DataFrame.
topics = pandas.read_csv('AOL-treated-Citizen-Lab-Classification-domain-match.csv', low_memory=False, on_bad_lines='warn', index_col=0)

CPU times: user 10.3 ms, sys: 337 µs, total: 10.6 ms
Wall time: 9.15 ms


In [4]:
display(data)

RandID            QueryTime               Domain
0              0  2006-05-13 00:12:16           edzapp.com
1              0  2006-05-13 00:15:28  riverdale.k12.or.us
2              0  2006-05-13 00:15:28  riverdale.k12.or.us
3              0  2006-05-13 00:22:39    salkeiz.k12.or.us
4              0  2006-05-13 00:22:39     greatschools.net
...          ...                  ...                  ...
19426288  521691  2006-03-07 21:42:51              ups.com
19426289  521691  2006-03-15 19:24:17           google.com
19426290  521691  2006-04-01 18:45:41           google.com
19426291  521691  2006-04-02 01:33:52       supercross.com
19426292  521691  2006-05-15 00:03:49         976-tuna.com

[19426293 rows x 3 columns]

In [5]:
display(topics)

domain                match  topics
0                   pchome.com.tw        pchome.com.tw  'COMM'
1                      fbc.com.my           fbc.com.my   'REL'
2                  proceso.com.mx       proceso.com.mx  'NEWS'
3                      yam.org.tw           yam.org.tw  'HUMR'
4             radioformula.com.mx  radioformula.com.mx  'NEWS'
...                           ...                  ...     ...
4964         arinternet.pr.gov.br               gov.br   'REL'
4965          seguranca.sp.gov.br               gov.br   'REL'
4966  probation.homeoffice.gov.uk               gov.uk   'REL'
4967             polmil.sp.gov.br               gov.br   'REL'
4968        portoalegre.rs.gov.br               gov.br   'REL'

[4969 rows x 3 columns]

### Statistics

In [6]:
%%time
data.info(verbose=True, memory_usage='deep', show_counts=True)

<class 'pandas.core.frame.DataFrame'>
Index: 19426293 entries, 0 to 19426292
Data columns (total 3 columns):
 #   Column     Non-Null Count     Dtype 
---  ------     --------------     ----- 
 0   RandID     19426293 non-null  int64 
 1   QueryTime  19426293 non-null  object
 2   Domain     19426293 non-null  object
dtypes: int64(1), object(2)
memory usage: 2.9 GB
CPU times: user 13.2 s, sys: 128 ms, total: 13.3 s
Wall time: 13.1 s


Number of unique values per attribute.

In [7]:
%%time
display(data.nunique())

RandID        521607
QueryTime    5469196
Domain       1300484
dtype: int64

CPU times: user 14 s, sys: 170 ms, total: 14.2 s
Wall time: 14.1 s


`RandID` statistics.

In [8]:
%%time
display(data['RandID'].value_counts().describe([.25,.50,.75,.80,.85,.90,.95,.96,.97,.98,.99]))

count    521607.000000
mean         37.243160
std         226.665095
min           1.000000
25%           3.000000
50%          10.000000
75%          33.000000
80%          44.000000
85%          62.000000
90%          92.000000
95%         160.000000
96%         186.000000
97%         224.000000
98%         284.000000
99%         404.000000
max      150802.000000
Name: count, dtype: float64

CPU times: user 302 ms, sys: 137 ms, total: 440 ms
Wall time: 428 ms


Number of `RandID`s per count of records, e.g. 1 `RandID` has 150802 records and 70406 `RandID`s have 1 record.

In [9]:
%%time
with pandas.option_context('display.max_rows', None):
    display(data['RandID'].value_counts().value_counts().sort_index(ascending=False).sort_index(ascending=False))

count
150802        1
6227          1
5741          1
4272          1
3775          1
3516          1
3445          1
3265          1
3238          1
3074          1
2954          1
2887          1
2879          1
2874          1
2791          1
2786          1
2752          1
2712          1
2697          1
2696          1
2645          1
2579          1
2571          1
2508          1
2466          1
2464          1
2410          1
2354          1
2325          1
2311          1
2303          1
2298          1
2297          1
2265          1
2237          1
2236          1
2212          1
2192          1
2189          1
2187          1
2180          1
2165          1
2145          1
2129          1
2089          1
2075          1
2072          1
2036          1
2026          1
2025          1
2021          1
2002          1
2000          1
1995          1
1994          1
1981          1
1975          1
1925          1
1877          1
1874          1
1839          1
1827          1
18

CPU times: user 332 ms, sys: 137 ms, total: 469 ms
Wall time: 458 ms


In [10]:
%%time
display(data[['RandID','Domain']].groupby(['RandID','Domain']).value_counts().max())

2997

CPU times: user 19.9 s, sys: 1.59 s, total: 21.5 s
Wall time: 21.5 s


Date and time range.

In [11]:
display(data['QueryTime'].min())

'2006-03-01 00:01:04'

In [12]:
display(data['QueryTime'].max())

'2006-05-31 23:59:59'

Number of unique `Domain` values.

In [13]:
%%time
data_domains_counts = dict(Counter(data['Domain'].to_list()))
display(len(data_domains_counts))

1300484

CPU times: user 5.6 s, sys: 96.7 ms, total: 5.69 s
Wall time: 5.55 s


Top `Domain` values by number of records.

In [14]:
top = 100
with pandas.option_context('display.max_rows', top, 'display.min_rows', top):
    display(pandas.Series(data_domains_counts).sort_values(ascending=False).head(top))

yahoo.com                438816
google.com               404663
myspace.com              228687
about.com                211395
ebay.com                 162841
wikipedia.org            125343
amazon.com               106647
msn.com                  104408
imdb.com                 103844
mapquest.com             101887
go.com                    84490
aol.com                   66069
craigslist.org            65153
nih.gov                   49781
bankofamerica.com         49362
geocities.com             41663
citysearch.com            41372
ask.com                   39935
hotmail.com               38392
bizrate.com               34706
tripadvisor.com           32624
tripod.com                32334
superpages.com            31453
nextag.com                27970
answers.com               26997
southwest.com             26096
microsoft.com             25028
azlyrics.com              24987
ca.gov                    24694
tv.com                    23474
irs.gov                   23378
cnn.com 

In [15]:
%%time
display(topics.nunique())

domain    4969
match     3586
topics     217
dtype: int64

CPU times: user 6.84 ms, sys: 194 µs, total: 7.03 ms
Wall time: 5.93 ms


In [16]:
%%time
all_topics = []
for tup in topics.itertuples():
    all_topics.extend(tup.topics.split(", "))

CPU times: user 5.55 ms, sys: 0 ns, total: 5.55 ms
Wall time: 5.48 ms


In [17]:
# Computes the number of unique seen topics.
display(len(set(all_topics)))
with pandas.option_context('display.max_rows', None):
    display(set(all_topics))

31

{"'ALDR'",
 "'ANON'",
 "'COMM'",
 "'COMT'",
 "'CTRL'",
 "'CULTR'",
 "'DATE'",
 "'ECON'",
 "'ENV'",
 "'FILE'",
 "'GAME'",
 "'GMB'",
 "'GOVT'",
 "'GRP'",
 "'HACK'",
 "'HATE'",
 "'HOST'",
 "'HUMR'",
 "'IGO'",
 "'LGBT'",
 "'MILX'",
 "'MISC'",
 "'MMED'",
 "'NEWS'",
 "'POLR'",
 "'PORN'",
 "'PROV'",
 "'PUBH'",
 "'REL'",
 "'SRCH'",
 "'XED'"}

## Reduction

In [18]:
%%time
# Inner merge of data and topics DataFrames.
df = pandas.merge(data, topics, how='inner', left_on='Domain', right_on='domain', suffixes=['_AOL', '_Citizen_Lab'])

CPU times: user 6.85 s, sys: 349 ms, total: 7.2 s
Wall time: 7.11 s


In [19]:
display(df)

RandID            QueryTime                  Domain   
0             1  2006-03-01 11:54:19          kidshealth.org  \
1             1  2006-03-01 11:54:19          kidshealth.org   
2             1  2006-03-01 11:54:19          kidshealth.org   
3             3  2006-05-08 18:24:29          kidshealth.org   
4             3  2006-05-08 18:39:34          kidshealth.org   
...         ...                  ...                     ...   
3135265  520871  2006-04-10 18:23:43              persga.org   
3135266  521630  2006-03-19 07:20:42            ibama.gov.br   
3135267  521630  2006-03-19 07:20:42  meioambiente.mg.gov.br   
3135268  521630  2006-04-07 00:47:52           woking.gov.uk   
3135269  521630  2006-04-07 00:49:40       flintshire.gov.uk   

                         domain           match  topics  
0                kidshealth.org  kidshealth.org  'PUBH'  
1                kidshealth.org  kidshealth.org  'PUBH'  
2                kidshealth.org  kidshealth.org  'PUBH'  
3                kidshealth.org  kidshealth.org  'PUBH'  
4                kidshealth.org  kidshealth.org  'PUBH'  
...                         ...             ...     ...  
3135265              persga.org      persga.org   'ENV'  
3135266            ibama.gov.br          gov.br   'REL'  
3135267  meioambiente.mg.gov.br          gov.br   'REL'  
3135268           woking.gov.uk          gov.uk   'REL'  
3135269       flintshire.gov.uk          gov.uk   'REL'  

[3135270 rows x 6 columns]

In [20]:
%%time
df = df.drop(columns=['domain','match'])

CPU times: user 301 ms, sys: 8 ms, total: 309 ms
Wall time: 306 ms


In [21]:
df = df.sort_values(by=['RandID', 'QueryTime'], ignore_index=True)

In [22]:
display(df)

RandID            QueryTime          Domain   
0             1  2006-03-01 11:54:19  kidshealth.org  \
1             1  2006-03-01 11:54:19  kidshealth.org   
2             1  2006-03-01 11:54:19  kidshealth.org   
3             1  2006-03-01 12:10:38     harvard.edu   
4             1  2006-03-08 21:16:04     harvard.edu   
...         ...                  ...             ...   
3135265  521689  2006-05-05 01:09:28         nih.gov   
3135266  521689  2006-05-05 02:00:33     typepad.com   
3135267  521691  2006-03-04 11:36:38          go.com   
3135268  521691  2006-03-15 19:24:17      google.com   
3135269  521691  2006-04-01 18:45:41      google.com   

                                                    topics  
0                                                   'PUBH'  
1                                                   'PUBH'  
2                                                   'PUBH'  
3                           'NEWS', 'CTRL', 'HUMR', 'PUBH'  
4                           'NEWS', 'CTRL', 'HUMR', 'PUBH'  
...                                                    ...  
3135265                                             'GOVT'  
3135266                                     'HOST', 'NEWS'  
3135267                                             'NEWS'  
3135268  'COMT', 'MMED', 'CTRL', 'CULTR', 'NEWS', 'SRCH...  
3135269  'COMT', 'MMED', 'CTRL', 'CULTR', 'NEWS', 'SRCH...  

[3135270 rows x 4 columns]

### Statistics

In [23]:
%%time
df.info(verbose=True, memory_usage='deep', show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3135270 entries, 0 to 3135269
Data columns (total 4 columns):
 #   Column     Non-Null Count    Dtype 
---  ------     --------------    ----- 
 0   RandID     3135270 non-null  int64 
 1   QueryTime  3135270 non-null  object
 2   Domain     3135270 non-null  object
 3   topics     3135270 non-null  object
dtypes: int64(1), object(3)
memory usage: 727.2 MB
CPU times: user 2.39 s, sys: 14.6 ms, total: 2.4 s
Wall time: 2.39 s


Number of unique values per attribute.

In [24]:
%%time
display(df.nunique())

RandID        342003
QueryTime    2083662
Domain          4969
topics           217
dtype: int64

CPU times: user 2.79 s, sys: 3.76 ms, total: 2.79 s
Wall time: 2.79 s


`RandID` statistics.

In [25]:
%%time
display(df['RandID'].value_counts().describe([.25,.50,.75,.80,.85,.90,.95,.96,.97,.98,.99]))

count    342003.000000
mean          9.167376
std          45.999573
min           1.000000
25%           1.000000
50%           3.000000
75%           8.000000
80%          11.000000
85%          14.000000
90%          20.000000
95%          34.000000
96%          40.000000
97%          48.000000
98%          62.000000
99%          92.000000
max       23505.000000
Name: count, dtype: float64

CPU times: user 80.3 ms, sys: 4.02 ms, total: 84.4 ms
Wall time: 82.2 ms


Number of `RandID`s per count of records, e.g. 1 `RandID` has 23505 records and 93630 `RandID`s have 1 record.

In [26]:
%%time
with pandas.option_context('display.max_rows', None):
    display(df['RandID'].value_counts().value_counts().sort_index(ascending=False).sort_index(ascending=False))

count
23505        1
2516         1
1379         1
1167         1
1011         1
980          1
968          1
902          1
863          1
822          1
774          1
759          1
755          1
748          1
719          1
715          1
699          1
690          2
677          1
656          1
651          1
644          1
640          2
636          1
634          1
633          1
627          1
626          1
622          1
621          1
611          1
609          1
597          1
596          1
591          1
586          1
582          1
581          1
578          1
572          1
571          1
566          1
564          1
561          1
558          1
550          1
547          1
543          1
541          1
538          1
533          1
529          2
524          1
521          1
520          1
518          1
516          3
515          2
512          1
511          1
510          1
498          1
497          1
496          3
495          1
494          1
493 

CPU times: user 68.7 ms, sys: 25 µs, total: 68.7 ms
Wall time: 67.7 ms


In [27]:
%%time
display(df[['RandID','Domain']].groupby(['RandID','Domain']).value_counts().max())

2997

CPU times: user 1.01 s, sys: 24.1 ms, total: 1.03 s
Wall time: 1.03 s


Date and time range.

In [28]:
display(df['QueryTime'].min())

'2006-03-01 00:01:04'

In [29]:
display(df['QueryTime'].max())

'2006-05-31 23:59:58'

Number of unique `Domain` values.

In [30]:
%%time
df_domains_counts = dict(Counter(df['Domain'].to_list()))
display(len(df_domains_counts))

4969

CPU times: user 251 ms, sys: 321 µs, total: 251 ms
Wall time: 244 ms


Top `Domain` values by number of records.

In [31]:
top = 100
with pandas.option_context('display.max_rows', top, 'display.min_rows', top):
    display(pandas.Series(df_domains_counts).sort_values(ascending=False).head(top))

yahoo.com                 438816
google.com                404663
myspace.com               228687
ebay.com                  162841
wikipedia.org             125343
amazon.com                106647
msn.com                   104408
imdb.com                  103844
go.com                     84490
aol.com                    66069
craigslist.org             65153
nih.gov                    49781
bankofamerica.com          49362
ask.com                    39935
tripod.com                 32334
microsoft.com              25028
azlyrics.com               24987
cnn.com                    23025
walmart.com                22537
angelfire.com              21570
nytimes.com                20145
pogo.com                   17137
wellsfargo.com             15878
bbc.co.uk                  15617
gamespot.com               14506
ign.com                    13987
army.mil                   13915
nbc.com                    13884
reference.com              12629
xanga.com                  11866
sapo.pt   

In [32]:
all_topics = []
for tup in df.itertuples():
    all_topics.extend(tup.topics.split(", "))

In [33]:
# Computes the number of unique seen topics.
display(len(set(all_topics)))
display(set(all_topics))

31

{"'ALDR'",
 "'ANON'",
 "'COMM'",
 "'COMT'",
 "'CTRL'",
 "'CULTR'",
 "'DATE'",
 "'ECON'",
 "'ENV'",
 "'FILE'",
 "'GAME'",
 "'GMB'",
 "'GOVT'",
 "'GRP'",
 "'HACK'",
 "'HATE'",
 "'HOST'",
 "'HUMR'",
 "'IGO'",
 "'LGBT'",
 "'MILX'",
 "'MISC'",
 "'MMED'",
 "'NEWS'",
 "'POLR'",
 "'PORN'",
 "'PROV'",
 "'PUBH'",
 "'REL'",
 "'SRCH'",
 "'XED'"}

## Save to file

In [34]:
df.to_csv('AOL-reduced-Citizen-Lab-Classification.csv')